### ISSUE TITLE
evaluate accuracy

### ISSUE THREAD

**Issue by cyLi-Tiger** [mit-han-lab/lmquant#4]
Thanks for your great work!

I use lmquant to generate the checkpoints for Qwen1.5-72b-chat with 
`python -m lmquant.llm.run configs/llm.yaml configs/qoq/gchn.yaml --model-name path_to_my_checkpoints --smooth-xw-alpha 0.3 --smooth-xw-beta 0.7`  and use `run_e2e.sh` in qserve to generate tokens, but the result seems wrong.

Is there a way that I can do end to end inference without qserve and evaluate the accuracy of the w4a8kv4 algorithm alone?

**Comment by synxlin** [comment]
Hi @cyLi-Tiger ,  if you would like to evaluate the accuracy alone, you can just use `lmquant` to evaluate with the fake-quant model. To evaluate with the wikitext 2 perplexity, you can use command
```
python -m lmquant.llm.run configs/llm.yaml configs/qoq/gchn.yaml --model-name path_to_my_checkpoints --smooth-xw-alpha 0.3 --smooth-xw-beta 0.7
```
To evaluate with the zero-shot accuracy, you can use command
```
python -m lmquant.llm.run configs/llm.yaml configs/qoq/gchn.yaml --model-name path_to_my_checkpoints --smooth-xw-alpha 0.3 --smooth-xw-beta 0.7 --eval-evaluator lm_eval --eval-tasks zero-shot
```
which will automatically add [wikitext, hellaswag, piqa, winogrande, arc_easy, arc_challenge](https://github.com/mit-han-lab/lmquant/blob/main/lmquant/llm/eval.py#L51) to the evaluation tasks using `lm_eval`.

**Comment by cyLi-Tiger** [comment]
Thanks for your prompt reply! @synxlin 

Take ppl evaluation as an example in QoQ, [here](https://github.com/mit-han-lab/lmquant/blob/c5c10897da4957cba47dd798080f5b2e1321f474/lmquant/llm/eval.py#L233), kv cache is unused, right? The above scripts you provide can't evaluate the accuracy under kv4. Please correct me if I miss something in your code, thank you again!

**Comment by synxlin** [comment]
Hi @cyLi-Tiger. 
- We replace the `apply_rotary_pos_emb` function in Attention forward [here](https://github.com/mit-han-lab/lmquant/blob/main/lmquant/llm/nn/attention.py#L147) and register the quantization hook to quantize KV cache [here](https://github.com/mit-han-lab/lmquant/blob/main/lmquant/quant/quantizer/base.py#L273). In the Huggingface `transformers` package, the KV cache will directly append/concat the results of `apply_rotary_pos_emb` for further computation (see [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L336) and [here](https://github.com/huggingface/transformers/blob/main/src/transformers/cache_utils.py#L331)).
- For wikitext-2 perplexity evaluation [here](https://github.com/mit-han-lab/lmquant/blob/c5c10897da4957cba47dd798080f5b2e1321f474/lmquant/llm/eval.py#L233), it is introduced by the prior work GPTQ (see [here](https://github.com/IST-DASLab/gptq/blob/main/llama.py#L213) and widely used in LLM quantization area. It only calculates the error rate *first* token.
- For per-channel W4A8 quantization, we directly extend unsigned INT4 with 4 zeros to become signed INT8 for INT8 computation. You can assume the quantization scale factor from INT4 to INT8 is `torch.ones`.
